File: apps/api/src/temporal.ts
import { Connection, Client } from '@temporalio/client';
import { randomUUID } from 'crypto';
import { encryptedDataConverter } from './temporal-data-converter';
import type { PipelineDefinition, DynamicWorkflowInput, DataRef, Environment } from '@dataflow/shared';
import { dataflowOpenLineageRunEvent } from '@dataflow/shared';
import { pool, withTenant } from './db';
import { consumeExecutionQuota, releaseExecutionQuota } from './services/usage';

// ─── Per-environment Temporal wiring ───────────────────────────────────────
// M3: each environment ('test' | 'prod') is its own Temporal namespace and task
// queue, served by its own worker pool (compose: worker-test / worker-prod).
// Clients share one Connection but are cached per namespace.
export function namespaceFor(env: Environment): string { return env; }
export function taskQueueFor(env: Environment): string { return `dynamic-dag-${env}`; }

let connection: Connection;
const clients = new Map<string, Client>();

export async function temporal(namespace = process.env.TEMPORAL_NAMESPACE ?? 'default'): Promise<Client> {
  let client = clients.get(namespace);
  if (!client) {
    if (!connection) {
      connection = await Connection.connect({ address: process.env.TEMPORAL_ADDRESS ?? 'localhost:7233' });
    }
    client = new Client({ connection, namespace, dataConverter: encryptedDataConverter });
    clients.set(namespace, client);
  }
  return client;
}

export async function fireExecution(
  def: PipelineDefinition, pipelineRowId: string,
  triggerType: string, env: Environment = 'test', payloadRef?: DataRef,
  encryptedDek?: string, retryOf?: string, backfillPartitionId?: string,
): Promise<string> {
  // Central enforcement so EVERY trigger path (manual, webhook, event) is
  // capped — not just manual runs gated by requireQuota. Throws before any
  // workflow starts, so an over-quota tenant never consumes a unit.
  await consumeExecutionQuota(def.tenantId);
  const executionId = `exec-${randomUUID()}`;
  const input: DynamicWorkflowInput = {
    definition: def,                                  // frozen — replay safe
    tenantId: def.tenantId,
    executionId,
    trigger: { type: triggerType, payloadRef, firedAt: new Date().toISOString() },
    encryptedDek,
  };
  const c = await temporal(namespaceFor(env));
  let handle: Awaited<ReturnType<typeof c.workflow.start>> | undefined;
  try {
    const started = await c.workflow.start('DynamicDAGWorkflow', {
      args: [input], taskQueue: taskQueueFor(env), workflowId: executionId,
    });
    handle = started;
    // Must run inside withTenant so the RLS policy on `executions` resolves correctly.
    await withTenant(def.tenantId, async client => {
      await client.query(`INSERT INTO executions
         (id, pipeline_id, tenant_id, trigger_type, build_id, environment, workflow_id, run_id, retry_of, backfill_partition_id)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`,
      [
        executionId, pipelineRowId, def.tenantId, triggerType,
        process.env.BUILD_ID ?? null, env, executionId, started.firstExecutionRunId, retryOf ?? null, backfillPartitionId ?? null,
      ]);
      if (backfillPartitionId) await client.query(
        `UPDATE backfill_partitions SET status='running',execution_id=$2 WHERE id=$1`,
        [backfillPartitionId, executionId],
      );
      if (process.env.OPENLINEAGE_URL) {
        const event = dataflowOpenLineageRunEvent({
          definition: def, pipelineKey: def.id, executionId, tenantId: def.tenantId,
          environment: env, phase: 'started', eventTime: new Date().toISOString(),
          namespace: process.env.OPENLINEAGE_NAMESPACE,
        });
        await client.query(
          `INSERT INTO openlineage_outbox (tenant_id,execution_id,event_type,payload)
           VALUES ($1,$2,'START',$3) ON CONFLICT (execution_id,event_type) DO NOTHING`,
          [def.tenantId, executionId, JSON.stringify(event)],
        );
      }
    });
    return executionId;
  } catch (error) {
    await handle?.terminate('execution metadata insert failed').catch(() => {});
    await releaseExecutionQuota(def.tenantId).catch(() => {});
    throw error;
  }
}

// ─── Cron trigger → Temporal Schedule (defined inside the pipeline itself) ───
// Schedules are env-scoped so a pipeline's test and prod crons run independently
// in their own namespaces.
export async function syncSchedule(def: PipelineDefinition, pipelineRowId: string, env: Environment = 'test') {
  // Only cron triggers need a Temporal Schedule; avoid connecting for manual/webhook/event.
  if (def.trigger.type !== 'cron') return;
  const c = await temporal(namespaceFor(env));
  const scheduleId = `sched-${def.id}-${env}`;
  // remove stale schedule, then recreate if the definition declares cron
  try { await (await c.schedule.getHandle(scheduleId)).delete(); } catch { /* none */ }
  await c.schedule.create({
    scheduleId,
    spec: { cronExpressions: [def.trigger.schedule] },
    action: {
      type: 'startWorkflow',
      workflowType: 'DynamicDAGWorkflow',
      taskQueue: taskQueueFor(env),
      args: [{
        definition: def, tenantId: def.tenantId,
        executionId: '',
        pipelineRowId,
        environment: env,
        trigger: { type: 'cron', firedAt: '' },
      }],
      workflowId: `schedule-${def.id}-${env}`,
      workflowIdReusePolicy: 'ALLOW_DUPLICATE',
    },
    policies: { overlap: 'SKIP' },  // never double-run the same pipeline
  } as any);
}

File: apps/workflow-go/internal/api/temporal.go
package api

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/dataflow-poc/workflow-go/internal/model"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	enumspb "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/client"
)

type quotaExceeded struct{ Used, Limit int }

func (e *quotaExceeded) Error() string { return "Quota exceeded" }

func monthStart() string { return time.Now().UTC().Format("2006-01") + "-01" }

func (s *Server) consumeQuota(ctx context.Context, tenantID string) error {
	return s.DB.TenantTx(ctx, tenantID, func(tx pgx.Tx) error {
		if _, err := tx.Exec(ctx, `INSERT INTO billing_plans (tenant_id) VALUES ($1) ON CONFLICT DO NOTHING`, tenantID); err != nil {
			return err
		}
		if _, err := tx.Exec(ctx, `INSERT INTO usage_counters (tenant_id,month,execution_count) VALUES ($1,$2,0) ON CONFLICT DO NOTHING`, tenantID, monthStart()); err != nil {
			return err
		}
		var limit, used int
		if err := tx.QueryRow(ctx, `SELECT bp.free_tier_limit+bp.extra_quota,uc.execution_count FROM billing_plans bp
      JOIN usage_counters uc ON uc.tenant_id=bp.tenant_id AND uc.month=$2 WHERE bp.tenant_id=$1 FOR UPDATE OF bp,uc`, tenantID, monthStart()).Scan(&limit, &used); err != nil {
			return err
		}
		if used >= limit {
			return &quotaExceeded{Used: used, Limit: limit}
		}
		_, err := tx.Exec(ctx, `UPDATE usage_counters SET execution_count=execution_count+1 WHERE tenant_id=$1 AND month=$2`, tenantID, monthStart())
		return err
	})
}

func (s *Server) releaseQuota(ctx context.Context, tenantID string) {
	_ = s.DB.TenantTx(ctx, tenantID, func(tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `UPDATE usage_counters SET execution_count=GREATEST(0,execution_count-1) WHERE tenant_id=$1 AND month=$2`, tenantID, monthStart())
		return err
	})
}

func (s *Server) fireExecution(ctx context.Context, def model.PipelineDefinition, pipelineRowID, triggerType string, environment model.Environment, payloadRef *model.DataRef, encryptedDEK, retryOf, partitionID string) (string, error) {
	if err := s.consumeQuota(ctx, def.TenantID); err != nil {
		return "", err
	}
	executionID := "exec-" + uuid.NewString()
	input := model.WorkflowInput{Definition: def, TenantID: def.TenantID, ExecutionID: executionID,
		Environment: environment, Trigger: model.TriggerInput{Type: triggerType, PayloadRef: payloadRef, FiredAt: time.Now().UTC().Format(time.RFC3339Nano)}, EncryptedDEK: encryptedDEK}
	temporalClient := s.Temporal[string(environment)]
	workflowName := "DynamicDAGWorkflow"
	if def.Execution != nil && def.Execution.Engine == "stream-direct" {
		workflowName = "StreamDirectWorkflow"
	}
	if def.Execution != nil && def.Execution.Engine == "spark-sql" {
		workflowName = "SparkJobWorkflow"
	}
	if def.Execution != nil && def.Execution.Engine == "flink-sql" {
		workflowName = "FlinkJobWorkflow"
	}
	run, err := temporalClient.ExecuteWorkflow(ctx, client.StartWorkflowOptions{ID: executionID, TaskQueue: "dynamic-dag-" + string(environment)}, workflowName, input)
	if err != nil {
		s.releaseQuota(ctx, def.TenantID)
		return "", err
	}
	err = s.DB.TenantTx(ctx, def.TenantID, func(tx pgx.Tx) error {
		_, err := tx.Exec(ctx, `INSERT INTO executions
      (id,pipeline_id,tenant_id,trigger_type,build_id,environment,workflow_id,run_id,retry_of,backfill_partition_id)
      VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`, executionID, pipelineRowID, def.TenantID, triggerType, nullString(os.Getenv("BUILD_ID")), environment, executionID, run.GetRunID(), nullString(retryOf), nullString(partitionID))
		if err != nil {
			return err
		}
		if partitionID != "" {
			_, err = tx.Exec(ctx, `UPDATE backfill_partitions SET status='running',execution_id=$2 WHERE id=$1`, partitionID, executionID)
		}
		return err
	})
	if err != nil {
		_ = temporalClient.TerminateWorkflow(ctx, executionID, run.GetRunID(), "execution metadata insert failed")
		s.releaseQuota(ctx, def.TenantID)
		return "", err
	}
	return executionID, nil
}

func (s *Server) syncSchedule(ctx context.Context, def model.PipelineDefinition, pipelineRowID string, environment model.Environment) error {
	if def.Trigger.Type != "cron" {
		return nil
	}
	temporalClient := s.Temporal[string(environment)]
	id := fmt.Sprintf("sched-%s-%s", def.ID, environment)
	_ = temporalClient.ScheduleClient().GetHandle(ctx, id).Delete(ctx)
	_, err := temporalClient.ScheduleClient().Create(ctx, client.ScheduleOptions{
		ID: id, Spec: client.ScheduleSpec{CronExpressions: []string{def.Trigger.Schedule}},
		Action: &client.ScheduleWorkflowAction{ID: fmt.Sprintf("schedule-%s-%s", def.ID, environment), Workflow: "DynamicDAGWorkflow", TaskQueue: "dynamic-dag-" + string(environment),
			Args: []interface{}{model.WorkflowInput{Definition: def, TenantID: def.TenantID, PipelineRowID: pipelineRowID, Environment: environment, Trigger: model.TriggerInput{Type: "cron"}}}},
		Overlap: enumspb.SCHEDULE_OVERLAP_POLICY_SKIP,
	})
	return err
}
